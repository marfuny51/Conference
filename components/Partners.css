.Full {
    color: white;
    text-align: center;
    font-family: 'Muli', sans-serif;
    font-size: 1.2vw;
    letter-spacing: 0.2vw;
}

.Partner {
    text-align: center;
    color: #00D970;
    margin: 2vw;
}

.Sort {
    font-size: 1.3vw;
    font-weight: bold;
    color: #a2eca3;
    letter-spacing: 0.1vw;
}

.Checkbox {
    width: 1.2vw;
    height: 1.2vw;
}

.NavLink {
    color: white;
}

.PartnerList {
    margin: 2vw;
}