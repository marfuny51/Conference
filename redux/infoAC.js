const INFO_LOADING='INFO_LOADING';
const INFO_ERROR='INFO_ERROR';
const INFO_SET='INFO_SET';

const infoLoadingAC=function() {
  return {
    type: INFO_LOADING,
  };
}

const infoErrorAC=function() {
  return {
    type: INFO_ERROR,
  };
}

const infoSetAC=function(info) {
  return {
    type: INFO_SET,
    info:info,
  };
}

export {
    infoLoadingAC,INFO_LOADING,
    infoErrorAC,INFO_ERROR,
    infoSetAC,INFO_SET,
}
