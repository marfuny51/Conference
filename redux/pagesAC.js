const PAGES_LOADING='PAGES_LOADING';
const PAGES_ERROR='PAGES_ERROR';
const PAGES_SET='PAGES_SET';

const pagesLoadingAC=function() {
  return {
    type: PAGES_LOADING,
  };
}

const pagesErrorAC=function() {
  return {
    type:PAGES_ERROR,
  };
}

const pagesSetAC=function(info) {
  return {
    type: PAGES_SET,
    info:info,
  };
}

export {
  pagesLoadingAC,PAGES_LOADING,
  pagesErrorAC,PAGES_ERROR,
  pagesSetAC,PAGES_SET,
}
