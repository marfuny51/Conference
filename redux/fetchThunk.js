import isoFetch from 'isomorphic-fetch';

import { infoLoadingAC, infoErrorAC, infoSetAC, infoSortAC } from "./infoAC";

function infoThunkAC(dispatch) {
    // Как и любой action creator, countriesThunkAC должен вернуть action,
    // только action будет не хэш, а ФУНКЦИЯ.
    // Все middleware стоят ДО редьюсеров, их задача - преобразовывать или фильтровать action-ы.
    // Конкретно middleware "thunk", если обнаруживает что action - функция а не хэш, 
    // ВЫПОЛНЯЕТ эту функцию и не пропускает её дальше, к редьюсерам.
    return function() {
        dispatch( infoLoadingAC() );
        isoFetch("http://localhost:3000/partners")
            .then( (response) => { // response - HTTP-ответ
                if (!response.ok) {
                    let Err=new Error("fetch error " + response.status);
                    Err.userMessage="Ошибка связи";
                    throw Err;
                }
                else
                    return response.json();
            })
            .then( (data) => {
                dispatch( infoSetAC(data) );
            })
            .catch( (error) => {
                console.error(error);
                dispatch( infoErrorAC() );
            })
        ;
    }

}

function infoThunkAC2(dispatch) {
    return function() {
        dispatch( infoLoadingAC() );
        isoFetch("http://localhost:3000/partners")
            .then( (response) => { // response - HTTP-ответ
                if (!response.ok) {
                    let Err=new Error("fetch error " + response.status);
                    Err.userMessage="Ошибка связи";
                    throw Err;
                }
                else
                    return response.json();
            })
            .then( (data) => {
                dispatch( infoSortAC(data) );
            })
            .catch( (error) => {
                console.error(error);
                dispatch( infoErrorAC() );
            })
        ;
    }

}

export {infoThunkAC};
export {infoThunkAC2};

