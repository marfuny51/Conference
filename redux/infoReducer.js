import { INFO_LOADING, INFO_ERROR, INFO_SET, INFO_SORT } from './infoAC';

const initState={

  status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
  data: null,

}

function infoReducer(state=initState,action) {
  switch (action.type) {

    case INFO_LOADING: {
      let newState={
        status:1,
        data:null,
      };
      return newState;
    }

    case INFO_ERROR: {
      let newState={
        status:2,
        data:null,
      };
      return newState;
    }

    case INFO_SET: {
      let newState={
        status:3,
        data:action.info,
      };
      return newState;
    }

    case INFO_SORT: {
      let newState={
        status:3,
        data:action.info.sort(function(a, b) {return (a.name > b.name) ? 1 : ((b.name > a.name) ? -1 : 0);})
      };
      console.log(action.info);
      return newState;
    }
    
    default:
      return state;
  }
}

export default infoReducer;
