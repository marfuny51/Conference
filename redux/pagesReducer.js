
import { PAGES_LOADING, PAGES_ERROR, PAGES_SET } from './pagesAC';

const initState={

  status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
  data: null,

}

function pagesReducer(state=initState,action) {
  switch (action.type) {

    case PAGES_LOADING: {
      let newState={
        status:1,
        data:null,
      };
      return newState;
    }

    case PAGES_ERROR: {
      let newState={
        status:2,
        data:null,
      };
      return newState;
    }

    case PAGES_SET: {
      let newState={
        status:3,
        data:JSON.parse(action.info.result),
      };
      return newState;
    }
    
    default:
      return state;
  }
}

export default pagesReducer;
