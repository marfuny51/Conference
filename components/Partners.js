import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import {connect} from 'react-redux';

import { infoThunkAC, infoThunkAC2 } from "../redux/fetchThunk";

import './Partners.css';

class Partners extends React.PureComponent {

  static propTypes = {
    info: PropTypes.object.isRequired,
  };

  state = {
    sort: false,
    line:'',
  }
            
  componentDidMount() {
    this.props.dispatch( infoThunkAC(this.props.dispatch) );
  }

  sort = () => {
    if (this.state.sort === false) {
      this.setState({sort: true});
      this.props.dispatch( infoThunkAC2(this.props.dispatch) );
    }
    else {
      this.setState({sort: false});
      this.props.dispatch( infoThunkAC(this.props.dispatch) );
    }
  }

  lineSearch = (EO) => {
    this.setState({line: EO.target.value.toString()}, this.search);
    //this.props.dispatch( infoThunkAC3(this.props.dispatch, this.state.line) );
  }

  search = () => {
    this.props.info.data = this.props.info.data.filter(partner => partner.name.indexOf(this.state.line)!=-1)
  }

  reset = () => {
    this.setState({sort: false, line: ''});
    this.props.dispatch( infoThunkAC(this.props.dispatch) );
  }

  render() {

    if ( this.props.info.status<=1 )
      return "загрузка...";

    if ( this.props.info.status===2 )
      return "ошибка загрузки данных";
    
    return (
      <div className='Partner'>
        <span className='Sort'>Sort information </span><input type='checkbox' className='Checkbox' checked= {this.state.sort} onChange={this.sort} /><br/>
        <input type='button' className='Reset' value='Reset' onClick={this.reset}/>
        <div> 
          {this.props.info.data.map(a=> 
          <p key={a.id}><NavLink to={"/partner/"+a.name} className="NavLink">{a.name}</NavLink></p>)
          }
        </div>
      </div>
    );

  }

}

const mapStateToProps = function (state) {
  return {
    info: state.info,
  };
};

export default connect(mapStateToProps)(Partners);
