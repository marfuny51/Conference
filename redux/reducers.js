import { combineReducers } from 'redux';

import buttonsReducer from "./buttonsReducer";

let combinedReducer=combineReducers({
    info: buttonsReducer,
});

export default combinedReducer;
