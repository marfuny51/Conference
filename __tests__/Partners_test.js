import React from 'react';
import Partners from '../components/Partners';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import combinedReducer from '../redux/reducers.js';
import { configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

let store=createStore(combinedReducer, applyMiddleware(thunk));

test('работа Partners', () => {

    let partners={"partners":[{"id": 1, "name": "MediaCorporation","country": "Germany"},{"id": 2,"name": "Sistem knowledges","country": "USA"}, {"id": 3,"name": "ELAM System", "country": "Belarus"}]};
    // создаём тестовую версию компонента
    const component = renderer.create(
        <Provider store={store}>
                  <Partners />
          </Provider>
    );
  
    // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
  
    wrapper.find('[htmlFor="checkbox"]').simulate('change', {
        target: { info: {partners:[{id: 3,name: "ELAM System", country: "Belarus"},{id: 1, name: "MediaCorporation",country: "Germany"}, {id: 2,name: "Sistem knowledges",country: "USA"}]} },
      });

    let componentTreeSort=component.toJSON();
    expect(componentTreeSort).toMatchSnapshot();
    
    wrapper.find('[htmlFor="checkbox"]').simulate('change', {
        target: { info: {partners:[{id: 1, name: "MediaCorporation",country: "Germany"},{id: 2,name: "Sistem knowledges",country: "USA"}, {id: 3,name: "ELAM System", country: "Belarus"}]} },
      });

    // получаем уже изменённый снэпшот
    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

    /*
    // можно эмулировать события, передавая в качестве объекта события то что нам нужно:
    wrapper.find('select').simulate('change', {
      target: { value: "hello" },
    });
    */
  
  });

  const mapStateToProps = function (state) {
    return {
      info: state.info,
    };
  };
  