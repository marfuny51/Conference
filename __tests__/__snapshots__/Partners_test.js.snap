// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа Partners 1`] = `"загрузка..."`;
